### PR TITLE
12.0 odoo_backup_sh bugfixes and doc fixes

### PR DESCRIPTION
**Pull request by em230418**

### CODE REVIEW

File: odoo_backup_sh/controllers/main.py [hunk 1]
```diff
@@ -131,6 +131,16 @@ def update_cloud_params(cls, cloud_params, redirect, call_from):
                 ).unlink()
         return cloud_params
 
+    @http.route('/web/database/backup', type='http', auth='none')
```
Comment (yelizariev) [inline odoo_backup_sh/controllers/main.py]: It cannot be auth=none here

File: odoo_backup_sh/controllers/main.py [hunk 2]
```diff
@@ -131,6 +131,16 @@ def update_cloud_params(cls, cloud_params, redirect, call_from):
...
+    def download_backup(self, backup_file_name):
+        headers = {
+            'Content-Disposition': 'attachment; filename="{0}"'.format(backup_file_name),
+            'Content-Type': 'application/octet-stream',
+        }
+        cloud_params = self.get_cloud_params(request.httprequest.url, call_from='frontend')
+        backup_object = BackupCloudStorage.get_object(cloud_params, backup_file_name)
+        return request.make_response(backup_object['Body'].read(), headers)
```
Comment (yelizariev) [inline odoo_backup_sh/controllers/main.py]: Looks like you allow to anyone download a backup

File: odoo_backup_sh/doc/index.rst [hunk 3]
```diff
@@ -75,3 +75,14 @@ Usage
 RESULT: Backup is restored in one click without any additional manipulations such as "downloading-uploading process".
 
 *The database restore may take some time before being ready*.
+
+Downloading backups
+===================
+
+* Proceed to the Backup list: ``/web/database/backups``
+* Click on ``Download`` button and wait until download completes
+* Backup is encrypted (it has `.enc` extension), extract it using `gpg` utility. As example:
```
Comment (yelizariev) [inline odoo_backup_sh/doc/index.rst]: ``If`` is missed?

File: odoo_backup_sh/doc/index.rst [hunk 4]
```diff
@@ -75,3 +75,14 @@ Usage
 RESULT: Backup is restored in one click without any additional manipulations such as "downloading-uploading process".
 
 *The database restore may take some time before being ready*.
+
+Downloading backups
+===================
+
+* Proceed to the Backup list: ``/web/database/backups``
```
Comment (yelizariev) [inline odoo_backup_sh/doc/index.rst]: Downloading must be done in backend

File: odoo_backup_sh_dropbox/doc/index.rst [hunk 5]
```diff
@@ -68,3 +68,15 @@ Usage
...
 RESULT: Backup is stored in the Dropbox cloud.
 *Note: Restore without additional downloading is available only for* `Odoo backup service <https://apps.odoo.com/apps/modules/12.0/odoo_backup_sh/>`__
+
+
+Downloading backups
+===================
+
+* Proceed to the Backup list: ``/web/database/backups``
+* Click on ``Download`` button and wait until download completes
+* Backup is encrypted (it has `.enc` extension), extract it using `gpg` utility. As example:
```
Comment (yelizariev) [inline odoo_backup_sh_dropbox/doc/index.rst]: In dropbox you can download directly from dropbox.

File: odoo_backup_sh/doc/index.rst [hunk 6]
```diff
@@ -75,3 +75,14 @@ Usage
...
+* Click on ``Download`` button and wait until download completes
+* Backup is encrypted (it has `.enc` extension), extract it using `gpg` utility. As example:
+
+```
+gpg --output OUTPUT_FILENAME.zip --decrypt INPUT_FILENAME.zip.enc
+```
```
Comment (yelizariev) [inline odoo_backup_sh/doc/index.rst]: Add a note like "Type the password when it will be prompted"

File: odoo_backup_sh/doc/index.rst (hunk 3 above)
Comment (Mirgalimova) [reply odoo_backup_sh/doc/index.rst]: fixed

File: odoo_backup_sh/doc/index.rst (hunk 4 above)
Comment (Mirgalimova) [reply odoo_backup_sh/doc/index.rst]: @em230418 can you fix it?

File: odoo_backup_sh/doc/index.rst (hunk 6 above)
Comment (Mirgalimova) [reply odoo_backup_sh/doc/index.rst]: fixed

File: odoo_backup_sh/doc/index.rst (hunk 4 above)
Comment (Mirgalimova) [reply odoo_backup_sh/doc/index.rst]: @em230418 please check)

**Comment by em230418** [comment]
@Mirgalimova @yelizariev I removed downloading notes from odoo_backup_sh_dropbox and odoo_backup_sh_google_disk and changed note in odoo_backup_sh. It is planned, that instructions given in last one will work for 2 others.

File: odoo_backup_sh/controllers/main.py [hunk 7]
```diff
@@ -131,6 +131,24 @@ def update_cloud_params(cls, cloud_params, redirect, call_from):
...
+        backup_filename = backup_info_record.backup_filename
+        cloud_params = self.get_cloud_params(request.httprequest.url, call_from='frontend')
+        backup_object = BackupCloudStorage.get_object(cloud_params, backup_filename)
+        return http.Response(
+            backup_object['Body'].iter_chunks(),
+            headers={
+                'Content-Disposition': 'attachment; filename="{0}"'.format(backup_filename),
+                'Content-Type': 'application/octet-stream',
+            }
+        )
```
[the remark this answers is not in the document]
Comment (em230418) [reply odoo_backup_sh/controllers/main.py]: > DId you try to open the link with demo user, for example?

Yes.
FireShot Capture 13 - db - Odoo

Should I add admin check in this function and similar functions?

[the remark this answers is not in the document]
Comment (yelizariev) [reply odoo_backup_sh/controllers/main.py]: We need to create new Access Groups and check them before providing the backup

File: odoo_backup_sh/security/security.xml [hunk 8]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0"?>
+<odoo>
+    <record id="base.group_user" model="res.groups">
+        <field name="implied_ids" eval="[(4, ref('odoo_backup_sh.group_user'))]"/>
```
Comment (yelizariev) [inline odoo_backup_sh/security/security.xml]: Wrong. You allow to see backups for all employee

File: odoo_backup_sh/models/odoo_backup_sh.py [hunk 9]
```diff
@@ -36,6 +36,32 @@
...
+        database,
+        datetime.strftime(upload_datetime, REMOTE_STORAGE_DATETIME_FORMAT)
+        if type(upload_datetime) is not str else upload_datetime,
+    )
+
+
+def get_backup_by_id(env, backup_id):
+    backup = env['odoo_backup_sh.backup_info'].search([
+        ('id', '=', backup_id),
+    ])
```
Comment (yelizariev) [inline odoo_backup_sh/models/odoo_backup_sh.py]: It's the same as 
`` env['odoo_backup_sh.backup_info'].browse(backup_id)
``

File: odoo_backup_sh/views/odoo_backup_sh_views.xml [hunk 10]
```diff
@@ -91,6 +100,7 @@ License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html).
                         <field name="encrypted"/>
                         <field name="backup_size" widget="float"/>
                     </group>
+                    <button class="btn btn-dark" type="action" name="%(download_backup)d" groups="odoo_backup_sh.group_manager">Download backup</button>
```
Comment (yelizariev) [inline odoo_backup_sh/views/odoo_backup_sh_views.xml]: Use more classic button please. Check example in built-in forms

Comment (yelizariev) [reply odoo_backup_sh/views/odoo_backup_sh_views.xml]: Also, the button has to be hidden for simulation records

**Comment by em230418** [comment]
> I'm not sure that this is correct. `len(backup)` probably returns len(ids), which is always 1 here, while the record may not exist.

https://github.com/odoo/odoo/blob/94a2c54fb50e39e3343f3183ead2121940ed0878/odoo/models.py#L4980-L4982

It does not always return 1. At least in 12.0

**Pull request merged**